workflow/util/util.go

```diff
@@ -853,178 +1218,152 @@ func FormulateRetryWorkflow(ctx context.Context, wf *wfv1.Workflow, restartSucce
-							}
-							if descendantNode.Type == wfv1.NodeTypePod {
-								newWF, resetParentGroupNodes = resetConnectedParentGroupNodes(wf, newWF, node, resetParentGroupNodes)
-								deletedPods, podsToDelete = deletePodNodeDuringRetryWorkflow(wf, *descendantNode, deletedPods, podsToDelete)
-								log.Debugf("Deleted pod node %s since it belongs to node %s", descendantNode.Name, node.Name)
-							}
-						}
-					} else {
-						log.Debugf("Reset non-pod/suspend/skipped node %s", node.Name)
-						newNode := node.DeepCopy()
-						newWF.Status.Nodes.Set(newNode.ID, resetNode(*newNode))
-					}
-				}
-			} else {
-				if !containsNode(resetParentGroupNodes, node.ID) {
-					log.Debugf("Node %s remains as is", node.Name)
-					newWF.Status.Nodes.Set(node.ID, node)
+	for k := range toDelete {
+		n := wf.Status.Nodes[k]
+		if n.Type == wfv1.NodeTypePod {
+			deletedPods, podsToDelete = deletePodNodeDuringRetryWorkflow(wf, n, deletedPods, podsToDelete)
+		}
+	}
+
+	for id, n := range wf.Status.Nodes {
+		shouldDelete := toDelete[id] || strings.HasPrefix(n.Name, onExitNodeName)
+		if _, err := newWf.Status.Nodes.Get(id); err != nil && !shouldDelete {
+			newWf.Status.Nodes.Set(id, *n.DeepCopy())
+		}
+		if n.Name == onExitNodeName {
+			queue := list.New()
+			queue.PushBack(n)
+			for {
+				currNode := queue.Front()
+				if currNode == nil {
+					break
 				}
-			}
-		case wfv1.NodeError, wfv1.NodeFailed, wfv1.NodeOmitted:
-			if isGroupNode(node) {
-				newNode := node.DeepCopy()
-				newWF.Status.Nodes.Set(newNode.ID, resetNode(*newNode))
-				log.Debugf("Reset %s node %s since it's a group node", node.Name, string(node.Phase))
-				continue
-			} else {
-				if node.Type != wfv1.NodeTypeRetry && isDescendantNodeSucceeded(wf, node, nodeIDsToReset) {
-					log.Debugf("Node %s remains as is since it has succeed child nodes.", node.Name)
-					newWF.Status.Nodes.Set(node.ID, node)
-					continue
+				curr := currNode.Value.(wfv1.NodeStatus)
+				deletedPods, podsToDelete = deletePodNodeDuringRetryWorkflow(wf, curr, deletedPods, podsToDelete)
+				for i := range curr.Children {
+					child, err := wf.Status.Nodes.Get(curr.Children[i])
+					if err != nil {
+						return nil, nil, err
+					}
+					queue.PushBack(child)
 				}
-				log.Debugf("Deleted %s node %s since it's not a group node", node.Name, string(node.Phase))
-				deletedPods, podsToDelete = deletePodNodeDuringRetryWorkflow(wf, node, deletedPods, podsToDelete)
-				log.Debugf("Deleted pod node: %s", node.Name)
-				deletedNodes[node.ID] = true
+				queue.Remove(currNode)
 			}
-			// do not add this status to the node. pretend as if this node never existed.
-		default:
-			// Do not allow retry of workflows with pods in Running/Pending phase
-			return nil, nil, errors.InternalErrorf("Workflow cannot be retried with node %s in %s phase", node.Name, node.Phase)
+		}
+		if n.Name == wf.Name && !shouldRetryFailedType(n.Type) {
+			newWf.Status.Nodes.Set(id, resetNode(*n.DeepCopy()))
```

Comment:
Do we need this `DeepCopy`, it seems to me that `resetNode` will cause a copy?

Reply:
It will cause a copy but only a shallow one. This isn't great for nodes that will mutate over time. 
A reset node will mutate so it does need a deep copy.  There probably isn't pointers shared across multiple nodes, but its better to be safe than introduce some weird aliasing bug. 

I was actually missing one so this comment helped me catch it, so thanks.